### DISCUSSION TITLE
Different behavior of jq bash and zsh.

### DISCUSSION THREAD

**Discussion by GLADtr** [stedolan/jq#3193] [General]
Hello everyone. 
Different behavior of jq bash and zsh.
Tell me, how can I bring them to one style of work? 
Below is an example:

```sh
╭─zsh in ~/ on master ✘ (origin/master)
╰$ docker image inspect nginx:latest | jq '.[].Config.ExposedPorts' 
{
  "80/tcp": {}
}
╭─zsh in ~/ on master ✘ (origin/master)(origin/master)
╰$ docker image inspect nginx:latest | jq .[].Config.ExposedPorts  
zsh: no matches found: .[].Config.ExposedPorts
╭─zsh~/ on master ✘ (origin/master)
╰$ bash                                                               
bash:~/$ docker image inspect nginx:latest | jq .[].Config.ExposedPorts
{
  "80/tcp": {}
}
bash:~/$ 
```

**Comment by wader** [comment]
Use single quotes in both cases? `jq '.[].Config.ExposedPorts'`. I guess `[]` makes zsh try to glob?

**Reply by GLADtr** [reply]
I showed all the working options that I found. It took me about an hour.

**Reply by GLADtr** [reply]
> Use single quotes in both cases? `jq '.[].Config.ExposedPorts'`. I guess `[]` makes zsh try to glob?

Sorry. I didn't mean to be rude. It's just all in the work process. Then I reread it and realized that I was very brief, bordering on rude. Let's continue the dialogue, please, if it brings benefits to the product and the community. I apologize again.

**Reply by wader** [reply]
No worries. Could you explain a bit more what the problem is? your looking for a single jq command that will work in both zsh and bash? if so i think you're best of single-quoting to no trigger any zsh or bash file name globbing etc

**Reply by GLADtr** [reply]
I discovered this by accident. In one of the books, the author used jq in the context of docker - I decided to try and I could not get a result similar to the author's until I started using single quotes. After rereading the author's text again, I realized that he was working with /bin/bash. My system is in zsh by default. I switched to bash and VOILA! Everything worked exactly as the author indicated. That is, depending on the shell, the formatting of the parameters changes the way the utility works. I decided to discuss this here - what if this is a bug? or is it a feature? An example of the difference in operation is in my original message... I assume that perhaps such a difference in the formatting of the parameters can cause incorrect operation in more complex cases of using jq. It is quite possible that I am wrong, in which case the topic can be closed. Thank you for not getting upset about my previous answer.

**Reply by wader** [reply]
Don't think it's a bug in jq, more that different shell environments handles argument splitting, escaping etc differently, and that is not really something jq can do anything about. There is a section in the jq manual about this https://jqlang.github.io/jq/manual/#invoking-jq (see the 3rd paragraph about quoting). btw if you do a `set +o nomatch` in zsh i think `jq .[].Config.ExposedPorts` (no quotes) will work also, unless there are filenames it matches of course. So in conclusion, in shells like bash, zsh etc always single quote.

**Reply by GLADtr** [reply]
I immediately applied your advice and it works! Thank you! It is not always enough to be attentive to catch the details when studying the documentation - I know from my practical experience - 99% of the answers to questions are already in the documentation. Thank you for your patience and informative conversation. Good luck in all your endeavors.
